Firestore/core/src/firebase/firestore/util/filesystem_posix.cc

```diff
@@ -130,51 +131,88 @@ Status DeleteFile(const Path& path) {
 }
 
 Status RecursivelyDeleteDir(const Path& parent) {
-  DIR* dir = ::opendir(parent.c_str());
-  if (!dir) {
-    if (errno == ENOENT) {
-      return Status::OK();
+  DirectoryIterator iter(parent);
+  for (; iter.Valid(); iter.Next()) {
+    Status status = RecursivelyDelete(iter.file());
+    if (!status.ok()) {
+      return status;
     }
-
-    return Status::FromErrno(errno, StringFormat("Could not read directory %s",
-                                                 parent.ToUtf8String()));
   }
+  if (!iter.status().ok()) {
+    return iter.status();
+  }
+  return DeleteDir(parent);
+}
 
-  Status result;
-  while (result.ok()) {
-    errno = 0;
-    struct dirent* entry = ::readdir(dir);
-    if (!entry) {
-      if (errno != 0) {
-        result = Status::FromErrno(
-            errno,
-            StringFormat("Could not read directory %s", parent.ToUtf8String()));
-      }
-      break;
-    }
+}  // namespace detail
 
-    if (::strcmp(".", entry->d_name) == 0 ||
-        ::strcmp("..", entry->d_name) == 0) {
-      continue;
-    }
+struct DirectoryIterator::Rep {
+  DIR* dir;
+  struct dirent* entry;
+};
 
-    Path child = parent.AppendUtf8(entry->d_name, strlen(entry->d_name));
-    result = RecursivelyDelete(child);
+DirectoryIterator::DirectoryIterator(
+    const firebase::firestore::util::Path& path)
+    : parent_(path), rep_(absl::make_unique<DirectoryIterator::Rep>()) {
+  rep_->dir = ::opendir(parent_.c_str());
+  if (!rep_->dir) {
+    if (errno == ENOENT) {
+      status_ = Status::OK();
+    } else {
+      status_ = Status::FromErrno(
+          errno,
+          StringFormat("Could not open directory %s", parent_.ToUtf8String()));
+    }
+  } else {
+    status_ = Status::OK();
+  }
+  if (status_.ok()) {
+    Advance();
   }
+}
 
-  if (::closedir(dir)) {
-    result.Update(Status::FromErrno(
-        errno,
-        StringFormat("Could not close directory %s", parent.ToUtf8String())));
+DirectoryIterator::~DirectoryIterator() {
+  if (rep_->dir) {
+    if (::closedir(rep_->dir) != 0) {
+      HARD_FAIL("Could not close directory %s", parent_.ToUtf8String());
+    }
   }
+}
 
-  if (result.ok()) {
-    result = DeleteDir(parent);
+void DirectoryIterator::Advance() {
+  HARD_ASSERT(status_.ok(), "Advancing an errored iterator");
+  errno = 0;
+  rep_->entry = ::readdir(rep_->dir);
+  if (!rep_->entry) {
+    if (errno != 0) {
+      status_ = Status::FromErrno(
+          errno, StringFormat("Could not read %s", parent_.ToUtf8String()));
+    }
+  } else if (status_.ok()) {
+    // Skip self- and parent-pointer
+    if (::strcmp(".", rep_->entry->d_name) == 0 ||
+        ::strcmp("..", rep_->entry->d_name) == 0) {
+      Advance();
+    }
   }
-  return result;
 }
 
-}  // namespace detail
+void DirectoryIterator::Next() {
+  HARD_ASSERT(Valid(), "Next() called on invalid iterator");
```

Comment:
It doesn't seem like there's a material difference between `Next` and `Advance`. Is it really worth keeping them separate?

Reply:
Yes, `Next()`, as part of the public API of the class, does an extra assert that would fail when the object is initially constructed: `rep_->entry != nullptr`.